lib/Service/OpenAiAPIService.php

```diff
@@ -85,6 +87,114 @@ public function getPromptHistory(string $userId, int $type): array {
 		return $this->promptMapper->getPromptsOfUser($userId, $type);
 	}
 
+	/**
+	 * Clear prompt history for a prompt type
+	 * @param string $userId
+	 * @param int $type
+	 */
+	public function clearPromptHistory(string $userId, int $type): void {
+		$this->promptMapper->deleteUserPromptsByType($userId, $type);
+	}
+
+	/**
+	 * @param string $userId
+	 */
+	private function hasOwnOpenAiApiKey(string $userId): bool {
+		if(!$this->isUsingOpenAi()){
+			return false;
+		}
+
+		if($this->config->getUserValue($userId, Application::APP_ID, 'api_key') !== ''){
+			return true;
+		}
+
+		return false;
+	}
+
+
+	/**
+	 * Check whether quota is exceeded for a user
+	 * @param string|null $userId
+	 * @param int $type
+	 * @return bool
+	 * @throws Exception
+	 */
+	public function isQuotaExceeded(?string $userId, int $type): bool {
+		if($userId === null){
+			return false;
+		}
+		
+		if(!array_key_exists($type, Application::QUOTA_TYPES)){
+			throw new Exception('Invalid quota type');
+		}
+
+		if($this->hasOwnOpenAiApiKey($userId)){
+			// User has specified own OpenAI API key, no quota limit:
+			return false;
+		}
+
+		// Get quota limits
+		$quota = intval(json_decode($this->config->getAppValue(Application::APP_ID, 'quotas', json_encode(Application::DEFAULT_QUOTAS)))[$type]->value);
+
+		if($quota === 0){
+			//  Unlimited quota:
+			return false;
+		}
+
+		$quotaPeriod = intval($this->config->getAppValue(Application::APP_ID, 'quota_period', Application::DEFAULT_QUOTA_PERIOD));
+		
+		try {
+			$quotaUsage= $this->quotaUsageMapper->getQuotaUnitsOfUserInTimePeriod($userId, $type, $quotaPeriod);
+		} catch (DoesNotExistException | MultipleObjectsReturnedException | Exception | RuntimeException $e) {
+			throw new Exception('Could not retrieve quota usage for user: ' . $userId . ' and quota type: ' . $type . '. Error: ' . $e->getMessage());
```

Comment:
You can pass the exception that triggered this catch to the new exception as a param to get a chained stack trace.

Reply:
Thanks, done!